src/mmda/utils/tools.py

```diff
@@ -1,36 +1,154 @@
             allocated_tokens.append(token)
         else:
             remaining_tokens.append(token)
     return allocated_tokens, remaining_tokens
 
 
+def box_groups_to_span_groups(
+        box_groups: List[BoxGroup], doc: Document, center: bool = True
+) -> List[SpanGroup]:
+    """Generate SpanGroups from BoxGroups.
+    Args
+        `box_groups` (List[BoxGroup])
+        `doc` (Document) base document annotated with pages, tokens, rows to
+        `center` (bool) defaults to True, considers tokens to be overlapping with boxes only if their centers overlap
+    Returns
+        List[SpanGroup] with each SpanGroup.spans corresponding to spans (sans boxes) of allocated tokens per box_group,
+        and each SpanGroup.box_group containing original box_groups
+    """
+    assert all([isinstance(group, BoxGroup) for group in box_groups])
+
+    all_page_tokens = dict()
+    avg_token_widths = dict()
+    derived_span_groups = []
+    token_box_in_box_group = None
+
+    for box_id, box_group in enumerate(box_groups):
+
+        all_tokens_overlapping_box_group = []
+
+        for box in box_group.boxes:
+
+            # Caching the page tokens to avoid duplicated search
+            if box.page not in all_page_tokens:
+                cur_page_tokens = all_page_tokens[box.page] = doc.pages[
+                    box.page
+                ].tokens
+                if token_box_in_box_group is None:
+                    # Determine whether box is stored on token SpanGroup span.box or in the box_group
+                    token_box_in_box_group = all(
+                        [
+                            (
+                                    (hasattr(token.box_group, "boxes") and len(token.box_group.boxes) == 1)
+                                    and token.spans[0].box is None
+                            )
+                            for token in cur_page_tokens
+                        ]
+                    )
+                    # Determine average width of tokens on this page
+                    if token_box_in_box_group and box.page not in avg_token_widths:
+                        avg_token_widths[box.page] = np.average([t.box_group.boxes[0].w for t in cur_page_tokens])
+                    elif not token_box_in_box_group and box.page not in avg_token_widths:
+                        avg_token_widths[box.page] = np.average([t.spans[0].box.w for t in cur_page_tokens])
+
+            else:
+                cur_page_tokens = all_page_tokens[box.page]
+
+            # Find all the tokens within the box
+            tokens_in_box, remaining_tokens = allocate_overlapping_tokens_for_box(
+                tokens=cur_page_tokens,
+                box=box,
+                token_box_in_box_group=token_box_in_box_group,
+                # pad a small amount so that extra narrow token boxes (when split at punctuation) are not missed
+                x=avg_token_widths.get(box.page, 0.0) * 0.5,
```

Comment:
If we want to maintain OG behavior for center=False, we should not pad in that case. Or make padding an option to this fn.